diff --git a/lib/c-stack.c b/lib/c-stack.c
old mode 100644
new mode 100755
index 9bbe6fe..9ac1c53
--- a/lib/c-stack.c
+++ b/lib/c-stack.c
@@ -52,7 +52,7 @@ typedef struct sigaltstack stack_t;
 #endif
 #ifndef SIGSTKSZ
 # define SIGSTKSZ 16384
-#elif HAVE_LIBSIGSEGV && SIGSTKSZ < 16384
+#elif (HAVE_LIBSIGSEGV && (SIGSTKSZ < 16384))
 /* libsigsegv 2.6 through 2.8 have a bug where some architectures use
    more than the Linux default of an 8k alternate stack when deciding
    if a fault was caused by stack overflow.  */
