diff -uprN kernel/arch/arm64/include/asm/hwcap.h kernel_new/arch/arm64/include/asm/hwcap.h
--- kernel/arch/arm64/include/asm/hwcap.h	2021-09-24 17:53:47.405317258 +0800
+++ kernel_new/arch/arm64/include/asm/hwcap.h	2021-09-27 14:45:46.614768036 +0800
@@ -127,5 +127,6 @@ enum {
 #endif
 };
 
+extern unsigned long elf_hwcap;
 #endif
 #endif
diff -uprN kernel/arch/arm64/kernel/cpufeature.c kernel_new/arch/arm64/kernel/cpufeature.c
--- kernel/arch/arm64/kernel/cpufeature.c	2021-09-24 17:53:47.445317211 +0800
+++ kernel_new/arch/arm64/kernel/cpufeature.c	2021-09-27 14:42:30.069467680 +0800
@@ -37,7 +37,7 @@
 #include <asm/virt.h>
 
 /* Kernel representation of AT_HWCAP and AT_HWCAP2 */
-static unsigned long elf_hwcap __read_mostly;
+unsigned long elf_hwcap __read_mostly;
 
 #ifdef CONFIG_COMPAT
 #define COMPAT_ELF_HWCAP_DEFAULT	\
diff -uprN kernel/drivers/iommu/arm-smmu-v3.c kernel_new/drivers/iommu/arm-smmu-v3.c
--- kernel/drivers/iommu/arm-smmu-v3.c	2021-09-24 17:53:51.785312088 +0800
+++ kernel_new/drivers/iommu/arm-smmu-v3.c	2021-09-27 14:42:13.739014805 +0800
@@ -209,7 +209,7 @@
 #define Q_BASE_RWA			(1UL << 62)
 #define Q_BASE_ADDR_MASK		GENMASK_ULL(51, 5)
 #define Q_BASE_LOG2SIZE			GENMASK(4, 0)
-#define Q_MAX_SZ_SHIFT			(PAGE_SHIFT + CONFIG_CMA_ALIGNMENT)
+#define Q_MAX_SZ_SHIFT			(PAGE_SHIFT + 8)
 
 /*
  * Stream table.
