--- a/libgcc/configure	2021-04-08 19:56:29.861762316 +0800
+++ b/libgcc/configure	2021-11-26 12:01:04.515038995 +0800
@@ -1060,7 +1060,7 @@ do
   -v | -verbose | --verbose | --verbos | --verbo | --verb)
     verbose=yes ;;
 
-  -version | --version | --versio | --versi | --vers | -V)
+  -version | --version | --versio | --versi | --vers)
     ac_init_version=: ;;
 
   -with-* | --with-*)
@@ -3530,7 +3530,7 @@ See \`config.log' for more details" "$LI
 $as_echo "$as_me:${as_lineno-$LINENO}: checking for C compiler version" >&5
 set X $ac_compile
 ac_compiler=$2
-for ac_option in --version -v -V -qversion; do
+for ac_option in --version -v ; do
   { { ac_try="$ac_compiler $ac_option >&5"
 case "(($ac_try" in
   *\"* | *\`* | *\\*) ac_try_echo=\$ac_try;;
@@ -6299,7 +6299,7 @@ do
   # Handling of the options.
   -recheck | --recheck | --rechec | --reche | --rech | --rec | --re | --r)
     ac_cs_recheck=: ;;
-  --version | --versio | --versi | --vers | --ver | --ve | --v | -V )
+  --version | --versio | --versi | --vers | --ver | --ve | --v )
     $as_echo "$ac_cs_version"; exit ;;
   --config | --confi | --conf | --con | --co | --c )
     $as_echo "$ac_cs_config"; exit ;;
