diff --git a/signal/signal.h b/signal/signal.h
old mode 100644
new mode 100755
index 5fd93382..6463c98e
--- a/signal/signal.h
+++ b/signal/signal.h
@@ -269,23 +269,10 @@ extern int sigwaitinfo (const sigset_t *__restrict __set,
 
    This function is a cancellation point and therefore not marked with
    __THROW.  */
-#  ifndef __USE_TIME_BITS64
 extern int sigtimedwait (const sigset_t *__restrict __set,
 			 siginfo_t *__restrict __info,
 			 const struct timespec *__restrict __timeout)
      __nonnull ((1));
-#  else
-#   ifdef __REDIRECT
-extern int __REDIRECT (sigtimedwait,
-                       (const sigset_t *__restrict __set,
-                        siginfo_t *__restrict __info,
-                        const struct timespec *__restrict __timeout),
-                       __sigtimedwait64)
-     __nonnull ((1));
-#   else
-#    define sigtimedwait __sigtimedwait64
-#   endif
-#  endif
 
 /* Send signal SIG to the process PID.  Associate data in VAL with the
    signal.  */
@@ -325,7 +312,6 @@ extern int siginterrupt (int __sig, int __interrupt) __THROW
   __attribute_deprecated_msg__ ("Use sigaction with SA_RESTART instead");
 
 # include <bits/sigstack.h>
-# include <bits/sigstksz.h>
 # include <bits/ss_flags.h>
 
 /* Alternate signal handler stack interface.
